packages/@react-aria/utils/src/domHelpers.ts

```diff
@@ -12,3 +12,83 @@ export const getOwnerWindow = (
   const doc = getOwnerDocument(el as Element | null | undefined);
   return doc.defaultView || window;
 };
+
+export const getRootNode = (
+  el: Element | null | undefined
+): Document | ShadowRoot | null => {
+  if (!el) {
+    // Return the main document if the element is null or undefined
+    return document;
+  }
+
+  // If the element is disconnected from the DOM, return null
+  if (!el.isConnected) {
+    return null;
+  }
+
+  // Get the root node of the element, or default to the document
+  const rootNode = el.getRootNode ? el.getRootNode() : document;
+
+  // Use nodeType to check the type of the rootNode
+  // We use nodeType instead of instanceof checks because instanceof fails across different
+  // contexts (e.g., iframes or windows), as each context has its own global objects and constructors.
+  // nodeType is a primitive value and is consistent across different contexts, making it
+  // reliable for cross-context type checking.
+  if (isDocument(rootNode)) {
+    // rootNode is a Document
+    return rootNode as Document;
+  }
+
+  if (isShadowRoot(rootNode)) {
+    // rootNode is a ShadowRoot (a specialized type of DocumentFragment)
+    // We check for the presence of the 'host' property to distinguish ShadowRoot from other DocumentFragments
+    return rootNode as ShadowRoot;
+  }
+
+  // For other types of nodes or DocumentFragments that are not ShadowRoots, return null
+  return null;
+};
+
+/**
+ * Retrieves a reference to the most appropriate "body" element for a given DOM context,
+ * accommodating both traditional DOM and Shadow DOM environments. When used with a Shadow DOM,
+ * it returns the body of the document to which the shadow root belongs, as shadow root is a document fragment,
+ * meaning that it doesn't have a body. When used with a regular document, it simply returns the document's body.
+ *
+ * @param {Document | ShadowRoot} root - The root document or shadow root from which to find the body.
+ * @returns {HTMLElement} - The "body" element of the document, or the document's body associated with the shadow root.
+ */
+export const getRootBody = (root: Document | ShadowRoot): HTMLElement => {
+  if (isShadowRoot(root)) {
+    return root.ownerDocument?.body;
+  } else {
+    return root.body;
+  }
+};
+
+/**
+ * Type guard that checks if a value is a Node. Verifies the presence and type of the nodeType property.
+ */
+function isNode(value: unknown): value is Node {
+  return value !== null &&
+    typeof value === 'object' &&
+    'nodeType' in value &&
+    typeof (value as Node).nodeType === 'number';
+}
+
+/**
+ * Type guard that checks if a node is a Document node. Uses nodeType for cross-context compatibility.
+ */
+export function isDocument(node: Node | null): node is Document {
+  return isNode(node) && node.nodeType === Node.DOCUMENT_NODE;
```

Comment:
Is the `isNode` check really necessary? You already know it's a node from the type declaration. TypeScript will ensure that.